wsmaster/codenvy-hosted-platform-api-impl/src/main/java/com/codenvy/api/workspace/LimitsCheckingWorkspaceManager.java

```diff
@@ -161,6 +186,35 @@ public WorkspaceImpl updateWorkspace(String id, Workspace update) throws Conflic
     }
 
     /**
+     * One of the checks in {@link #checkRamAndPropagateStart(WorkspaceConfig, String, String, String, WorkspaceCallback)}
+     * requests to docker for system ram amount. This request can take much time, so we can not mark the method as synchronized
+     * because this will reduce performance. The solution is to make the system ram check optimistic by limiting
+     * same time start workspace requests. All requests to start workspace performed at the same time can receive not actual
+     * system ram limit status, so limits.workspace.start.throughput property configures how many requests to handle at the same time.
+     * If limits.workspace.start.threads.limit property is set to 0 or less then 0 all requests will be handled at the same time.
+     */
+    private <T extends WorkspaceImpl> T checkRamAndPropagateLimitedStart(WorkspaceConfig config,
+                                                                         String envName,
+                                                                         String namespace,
+                                                                         String workspaceId,
+                                                                         WorkspaceCallback<T> callback) throws ServerException,
+                                                                                                               NotFoundException,
+                                                                                                               ConflictException {
+        if (startSemaphore == null) {
+            return checkRamAndPropagateStart(config, envName, namespace, workspaceId, callback);
+        } else {
+            try {
+                startSemaphore.acquire();
+                return checkRamAndPropagateStart(config, envName, namespace, workspaceId, callback);
+            } catch (InterruptedException e) {
+                throw new ServerException("Starting workspace was interrupted", e);
```

Comment:
it shouldn't happen normally, can you please just rethrow `InterruptedException` as `ServerException` with the same message like `throw new ServerException(e.getMessage(), e)`